src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1480,8 +1486,31 @@ private async Task<Solution> ComputeFrozenSolutionAsync(CancellationToken cancel
         /// </summary>
         internal Solution WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, CancellationToken cancellationToken)
         {
-            var newCompilationState = this.CompilationState.WithFrozenPartialCompilationIncludingSpecificDocument(documentId, cancellationToken);
-            return new Solution(newCompilationState);
+            AsyncLazy<Solution>? lazySolution;
+            lock (_documentIdToFrozenSolution)
+            {
+                if (!_documentIdToFrozenSolution.TryGetValue(documentId, out lazySolution))
+                {
+                    // in a local function to prevent lamdba allocations when not needed.
+                    lazySolution = CreateLazyFrozenSolution(this.CompilationState, documentId);
+                    _documentIdToFrozenSolution.Add(documentId, lazySolution);
+                }
+            }
```

Comment:
this should be cheap/fast, as we're not doing anything beyond looking up or creating the lazy (not the actual computation).